<a href='http://moncoachdata.com'><img src= 'https://moncoachdata.com/wp-content/uploads/MonCoachData-cours/MonCoachData.jpg' width=400/></a>


---


<center><em>Copyright MonCoachData (tous droits réservés)</em></center>
<center><em>Pour plus d'informations, visitez notre site <a href='http://moncoachdata.com'>moncoachdata.com</a></em></center>



---

# DataFrames

Tout au long du cours, la plupart de nos explorations de données se feront avec des DataFrames. Les DataFrames sont un outil extrêmement puissant et une extension naturelle de la série Pandas. Par définition :

**Un DataFrame Pandas est constitué de plusieurs séries Pandas qui partagent des valeurs d'index.**

## Importations

In [ ]:
import numpy as np
import pandas as pd

## Création d'un DataFrame à partir d'objets Python

In [ ]:
# help(pd.DataFrame)

In [ ]:
# Assurez-vous que l'appel du seed se trouve dans la même cellule que l'appel d'une fonction random
# https://stackoverflow.com/questions/21494489/what-does-numpy-random-seed0-do
np.random.seed(101)
mydata = np.random.randint(0,101,(4,3))

In [ ]:
mydata

array([[95, 11, 81],
       [70, 63, 87],
       [75,  9, 77],
       [40,  4, 63]])

In [ ]:
myindex = ['CA','NY','AZ','TX']

In [ ]:
mycolumns = ['Jan','Feb','Mar']

In [ ]:
df = pd.DataFrame(data=mydata)
df

0   1   2
0  95  11  81
1  70  63  87
2  75   9  77
3  40   4  63

In [ ]:
df = pd.DataFrame(data=mydata,index=myindex)
df

0   1   2
CA  95  11  81
NY  70  63  87
AZ  75   9  77
TX  40   4  63

In [ ]:
df = pd.DataFrame(data=mydata,index=myindex,columns=mycolumns)
df 

Jan  Feb  Mar
CA   95   11   81
NY   70   63   87
AZ   75    9   77
TX   40    4   63

In [ ]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 4 entries, CA to TX
Data columns (total 3 columns):
 #   Column  Non-Null Count  Dtype
---  ------  --------------  -----
 0   Jan     4 non-null      int64
 1   Feb     4 non-null      int64
 2   Mar     4 non-null      int64
dtypes: int64(3)
memory usage: 128.0+ bytes



## Lecture d'un fichier .csv pour un DataFrame

----

## Note : Nous examinerons toutes sortes d'entrées et de sorties de données - Data inputs et outputs (.html, .csv, .xlxs, etc...) plus tard dans le cours ! Pour l'instant, nous avons juste besoin de lire un simple fichier .csv.

----

Les fichiers CSV (Comma Separated Values = Valeurs séparées par des virgules) sont des fichiers texte qui utilisent des virgules comme délimiteurs de champ.

### Avec Google Colab

Importer un fichier directement depuis le menu de gauche (Fichiers). Puis, appeler directement le fichier avec :
    
        df = pd.read_csv('un_fichier.csv')
        
ou clic droit sur le nom du fichier uploadé / copier le chemin d'accès et coller le dans la méthode read_csv(' ici ')
    
        df = pd.read_csv('/content/example.csv')
        
**Attention : Il faudra toujours uploader le fichier de nouveau lorsque votre notebook sur Google Colab se recycle.**

ou Connecter (Google) Drive à Google Colab, cliquez sur 'Drive' puis sélectionner le fichier csv, clic droit / copier chemin d'accès et le coller.

    
        df = pd.read_csv('coller le chemin d'accès du fichier CSV depuis le Drive ici...')

### Avec Distribution Anaconda ou Autres

Les fichiers de valeurs séparées par des virgules sont des fichiers texte qui utilisent des virgules comme délimiteurs de champs.
Vous devrez peut-être installer <tt>xlrd</tt> et <tt>openpyxl</tt>.<br>
Dans votre terminal/commande (ou directement depuis jupyter notebook) :

    conda install xlrd
    conda install openpyxl

Ensuite, redémarrez Jupyter Notebook ou Spyder.
(ou utilisez pip install si vous n'utilisez pas la distribution Anaconda)

### Comprendre le chemin (path) des fichiers

Vous avez deux options lorsque vous lisez un dossier avec Pandas :

1. Si votre fichier .py ou votre notebook .ipynb se trouve dans le **même** dossier que le fichier .csv que vous voulez lire, il vous suffit d'indiquer le nom du fichier sous forme de chaîne de caractères (string), par exemple :
    
        df = pd.read_csv('un_fichier.csv')
        
2. Indiquez le chemin d'accès complet au fichier si vous vous trouvez dans un autre répertoire. Le chemin d'accès doit être correct à 100% pour que cela fonctionne. Par exemple sur Windows (sur Mac/Linux, ce seront des slashs simples '/') :

        df = pd.read_csv("C:\\Users\\nom_d'utilisateur\\files\\un_fichier.csv")

### Imprimer le chemin du fichier du répertoire actuel avec pwd

In [ ]:
pwd

'/content'

### Lister les fichiers dans votre répertoire actuel avec ls

In [ ]:
ls

sample_data/  tips.csv


In [ ]:
df = pd.read_csv('tips.csv')

In [ ]:
df

total_bill   tip     sex  ...          Payer Name         CC Number Payment ID
0         16.99  1.01  Female  ...  Christy Cunningham  3560325168603410    Sun2959
1         10.34  1.66    Male  ...      Douglas Tucker  4478071379779230    Sun4608
2         21.01  3.50    Male  ...      Travis Walters  6011812112971322    Sun4458
3         23.68  3.31    Male  ...    Nathaniel Harris  4676137647685994    Sun5260
4         24.59  3.61  Female  ...        Tonya Carter  4832732618637221    Sun2251
..          ...   ...     ...  ...                 ...               ...        ...
239       29.03  5.92    Male  ...       Michael Avila  5296068606052842    Sat2657
240       27.18  2.00  Female  ...      Monica Sanders  3506806155565404    Sat1766
241       22.67  2.00    Male  ...          Keith Wong  6011891618747196    Sat3880
242       17.82  1.75    Male  ...        Dennis Dixon     4375220550950      Sat17
243       18.78  3.00  Female  ...     Michelle Hardin  3511451626698139    Thur672

[244 rows x 11 columns]

----
A propos de ce DataSet (au cas où vous seriez intéressé)

* Description
    * Un serveur a enregistré des informations sur chaque pourboire qu'il a reçu sur une période de quelques mois dans un restaurant. Il a recueilli plusieurs variables :

* Format
    * Un dataFrame avec 244 lignes et 7 variables.

* Détails
    * **tip** : Pourboire en dollars,
    * **total_bill** : Facture en dollars,
    * **sex** : sexe du payeur de la facture,
    * **somker** : s'il y avait des fumeurs dans la fête,
    * **day** jour de la semaine,
    * **time** : l'heure de la journée,
    * **size** : taille de la fête.

Au total, il a enregistré 244 pourboires. Les données ont été rapportées dans une collection d'études de cas pour les statistiques commerciales (Bryant & Smith 1995).

* Références
    * Bryant, P. G. et Smith, M (1995) Practical Data Analysis : Case Studies in Business Statistics. Homewood, IL : Richard D. Irwin Publishing :
    
* Note : Nous avons créé quelques colonnes supplémentaires avec de **fausses données**, notamment le nom (**Payer Name**), le numéro de CC (**CC Number**) et l'ID de paiement (**Payment ID**).

----

## Obtenir des informations de base sur le DataFrame

In [ ]:
df.columns

Index(['total_bill', 'tip', 'sex', 'smoker', 'day', 'time', 'size',
       'price_per_person', 'Payer Name', 'CC Number', 'Payment ID'],
      dtype='object')

In [ ]:
df.index

RangeIndex(start=0, stop=244, step=1)

In [ ]:
df.head(3)

total_bill   tip     sex  ...          Payer Name         CC Number Payment ID
0       16.99  1.01  Female  ...  Christy Cunningham  3560325168603410    Sun2959
1       10.34  1.66    Male  ...      Douglas Tucker  4478071379779230    Sun4608
2       21.01  3.50    Male  ...      Travis Walters  6011812112971322    Sun4458

[3 rows x 11 columns]

In [ ]:
df.tail(3)

total_bill   tip     sex  ...       Payer Name         CC Number Payment ID
241       22.67  2.00    Male  ...       Keith Wong  6011891618747196    Sat3880
242       17.82  1.75    Male  ...     Dennis Dixon     4375220550950      Sat17
243       18.78  3.00  Female  ...  Michelle Hardin  3511451626698139    Thur672

[3 rows x 11 columns]

In [ ]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 244 entries, 0 to 243
Data columns (total 11 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   total_bill        244 non-null    float64
 1   tip               244 non-null    float64
 2   sex               244 non-null    object 
 3   smoker            244 non-null    object 
 4   day               244 non-null    object 
 5   time              244 non-null    object 
 6   size              244 non-null    int64  
 7   price_per_person  244 non-null    float64
 8   Payer Name        244 non-null    object 
 9   CC Number         244 non-null    int64  
 10  Payment ID        244 non-null    object 
dtypes: float64(3), int64(2), object(6)
memory usage: 21.1+ KB


In [ ]:
len(df)

244

In [ ]:
df.describe()

total_bill         tip        size  price_per_person     CC Number
count  244.000000  244.000000  244.000000        244.000000  2.440000e+02
mean    19.785943    2.998279    2.569672          7.888197  2.563496e+15
std      8.902412    1.383638    0.951100          2.914234  2.369340e+15
min      3.070000    1.000000    1.000000          2.880000  6.040679e+10
25%     13.347500    2.000000    2.000000          5.800000  3.040731e+13
50%     17.795000    2.900000    2.000000          7.255000  3.525318e+15
75%     24.127500    3.562500    3.000000          9.390000  4.553675e+15
max     50.810000   10.000000    6.000000         20.270000  6.596454e+15

In [ ]:
df.describe().transpose()

count          mean  ...           75%           max
total_bill        244.0  1.978594e+01  ...  2.412750e+01  5.081000e+01
tip               244.0  2.998279e+00  ...  3.562500e+00  1.000000e+01
size              244.0  2.569672e+00  ...  3.000000e+00  6.000000e+00
price_per_person  244.0  7.888197e+00  ...  9.390000e+00  2.027000e+01
CC Number         244.0  2.563496e+15  ...  4.553675e+15  6.596454e+15

[5 rows x 8 columns]

## Sélection et indexation

Apprenons à extraire des informations d'un DataFrame.

### Colonnes

Nous allons commencer à apprendre comment extraire des informations à partir des colonnes.

In [ ]:
df.head()

total_bill   tip     sex  ...          Payer Name         CC Number Payment ID
0       16.99  1.01  Female  ...  Christy Cunningham  3560325168603410    Sun2959
1       10.34  1.66    Male  ...      Douglas Tucker  4478071379779230    Sun4608
2       21.01  3.50    Male  ...      Travis Walters  6011812112971322    Sun4458
3       23.68  3.31    Male  ...    Nathaniel Harris  4676137647685994    Sun5260
4       24.59  3.61  Female  ...        Tonya Carter  4832732618637221    Sun2251

[5 rows x 11 columns]

#### Saisir une seule colonne

In [ ]:
df['to']

0      16.99
1      10.34
2      21.01
3      23.68
4      24.59
       ...  
239    29.03
240    27.18
241    22.67
242    17.82
243    18.78
Name: total_bill, Length: 244, dtype: float64

In [ ]:
type(df['total_bill'])

pandas.core.series.Series

#### Saisir plusieurs colonnes

In [ ]:
# Notez que c'est une liste python de noms de colonnes ! D'où les doubles crochets.
df[['total_bill','tip']]

total_bill   tip
0         16.99  1.01
1         10.34  1.66
2         21.01  3.50
3         23.68  3.31
4         24.59  3.61
..          ...   ...
239       29.03  5.92
240       27.18  2.00
241       22.67  2.00
242       17.82  1.75
243       18.78  3.00

[244 rows x 2 columns]

#### Créer de nouvelles colonnes

In [ ]:
df['tip_percentage'] = 100* df['tip'] / df['total_bill']

In [ ]:
df.head()

total_bill   tip     sex  ...         CC Number Payment ID tip_percentage
0       16.99  1.01  Female  ...  3560325168603410    Sun2959       5.944673
1       10.34  1.66    Male  ...  4478071379779230    Sun4608      16.054159
2       21.01  3.50    Male  ...  6011812112971322    Sun4458      16.658734
3       23.68  3.31    Male  ...  4676137647685994    Sun5260      13.978041
4       24.59  3.61  Female  ...  4832732618637221    Sun2251      14.680765

[5 rows x 12 columns]

In [ ]:
df['price_per_person'] = df['total_bill'] / df['size']

In [ ]:
df.head()

total_bill   tip     sex  ...         CC Number Payment ID tip_percentage
0       16.99  1.01  Female  ...  3560325168603410    Sun2959       5.944673
1       10.34  1.66    Male  ...  4478071379779230    Sun4608      16.054159
2       21.01  3.50    Male  ...  6011812112971322    Sun4458      16.658734
3       23.68  3.31    Male  ...  4676137647685994    Sun5260      13.978041
4       24.59  3.61  Female  ...  4832732618637221    Sun2251      14.680765

[5 rows x 12 columns]

In [ ]:
help(np.round)

Help on function round_ in module numpy:

round_(a, decimals=0, out=None)
    Round an array to the given number of decimals.
    
    See Also
    --------
    around : equivalent function; see for details.



#### Ajuster les colonnes existantes

In [ ]:
# Parce que les pandas sont basés sur numpy, nous obtenons des capacités incroyables avec les fonctions universelles de NumPy !
df['price_per_person'] = np.round(df['price_per_person'],2)

In [ ]:
df.head()

total_bill   tip     sex  ...         CC Number Payment ID tip_percentage
0       16.99  1.01  Female  ...  3560325168603410    Sun2959       5.944673
1       10.34  1.66    Male  ...  4478071379779230    Sun4608      16.054159
2       21.01  3.50    Male  ...  6011812112971322    Sun4458      16.658734
3       23.68  3.31    Male  ...  4676137647685994    Sun5260      13.978041
4       24.59  3.61  Female  ...  4832732618637221    Sun2251      14.680765

[5 rows x 12 columns]

#### Supprimer des colonnes

In [ ]:
# df.drop('tip_percentage',axis=1)

In [ ]:
df = df.drop("tip_percentage",axis=1)

In [ ]:
df.head()

total_bill   tip     sex  ...          Payer Name         CC Number Payment ID
0       16.99  1.01  Female  ...  Christy Cunningham  3560325168603410    Sun2959
1       10.34  1.66    Male  ...      Douglas Tucker  4478071379779230    Sun4608
2       21.01  3.50    Male  ...      Travis Walters  6011812112971322    Sun4458
3       23.68  3.31    Male  ...    Nathaniel Harris  4676137647685994    Sun5260
4       24.59  3.61  Female  ...        Tonya Carter  4832732618637221    Sun2251

[5 rows x 11 columns]

## Bases de l'index

Avant de passer en revue les mêmes tâches de récupération pour les lignes (rows), essayons de comprendre les bases de l'index d'un DataFrame Pandas.

In [ ]:
df.head()

total_bill   tip     sex  ...          Payer Name         CC Number Payment ID
0       16.99  1.01  Female  ...  Christy Cunningham  3560325168603410    Sun2959
1       10.34  1.66    Male  ...      Douglas Tucker  4478071379779230    Sun4608
2       21.01  3.50    Male  ...      Travis Walters  6011812112971322    Sun4458
3       23.68  3.31    Male  ...    Nathaniel Harris  4676137647685994    Sun5260
4       24.59  3.61  Female  ...        Tonya Carter  4832732618637221    Sun2251

[5 rows x 11 columns]

In [ ]:
df.index

RangeIndex(start=0, stop=244, step=1)

In [ ]:
df.set_index('Payment ID')

total_bill   tip  ...          Payer Name         CC Number
Payment ID                    ...                                      
Sun2959          16.99  1.01  ...  Christy Cunningham  3560325168603410
Sun4608          10.34  1.66  ...      Douglas Tucker  4478071379779230
Sun4458          21.01  3.50  ...      Travis Walters  6011812112971322
Sun5260          23.68  3.31  ...    Nathaniel Harris  4676137647685994
Sun2251          24.59  3.61  ...        Tonya Carter  4832732618637221
...                ...   ...  ...                 ...               ...
Sat2657          29.03  5.92  ...       Michael Avila  5296068606052842
Sat1766          27.18  2.00  ...      Monica Sanders  3506806155565404
Sat3880          22.67  2.00  ...          Keith Wong  6011891618747196
Sat17            17.82  1.75  ...        Dennis Dixon     4375220550950
Thur672          18.78  3.00  ...     Michelle Hardin  3511451626698139

[244 rows x 10 columns]

In [ ]:
df.head()

total_bill   tip     sex  ...          Payer Name         CC Number Payment ID
0       16.99  1.01  Female  ...  Christy Cunningham  3560325168603410    Sun2959
1       10.34  1.66    Male  ...      Douglas Tucker  4478071379779230    Sun4608
2       21.01  3.50    Male  ...      Travis Walters  6011812112971322    Sun4458
3       23.68  3.31    Male  ...    Nathaniel Harris  4676137647685994    Sun5260
4       24.59  3.61  Female  ...        Tonya Carter  4832732618637221    Sun2251

[5 rows x 11 columns]

In [ ]:
df = df.set_index('Payment ID')

In [ ]:
df.head()

total_bill   tip  ...          Payer Name         CC Number
Payment ID                    ...                                      
Sun2959          16.99  1.01  ...  Christy Cunningham  3560325168603410
Sun4608          10.34  1.66  ...      Douglas Tucker  4478071379779230
Sun4458          21.01  3.50  ...      Travis Walters  6011812112971322
Sun5260          23.68  3.31  ...    Nathaniel Harris  4676137647685994
Sun2251          24.59  3.61  ...        Tonya Carter  4832732618637221

[5 rows x 10 columns]

In [ ]:
df = df.reset_index()

In [ ]:
df.head()

Payment ID  total_bill  ...          Payer Name         CC Number
0    Sun2959       16.99  ...  Christy Cunningham  3560325168603410
1    Sun4608       10.34  ...      Douglas Tucker  4478071379779230
2    Sun4458       21.01  ...      Travis Walters  6011812112971322
3    Sun5260       23.68  ...    Nathaniel Harris  4676137647685994
4    Sun2251       24.59  ...        Tonya Carter  4832732618637221

[5 rows x 11 columns]

### Lignes (rows)

Explorons maintenant ces mêmes concepts mais avec des lignes.

In [ ]:
df.head()

Payment ID  total_bill  ...          Payer Name         CC Number
0    Sun2959       16.99  ...  Christy Cunningham  3560325168603410
1    Sun4608       10.34  ...      Douglas Tucker  4478071379779230
2    Sun4458       21.01  ...      Travis Walters  6011812112971322
3    Sun5260       23.68  ...    Nathaniel Harris  4676137647685994
4    Sun2251       24.59  ...        Tonya Carter  4832732618637221

[5 rows x 11 columns]

In [ ]:
df = df.set_index('Payment ID')

In [ ]:
df.head()

total_bill   tip  ...          Payer Name         CC Number
Payment ID                    ...                                      
Sun2959          16.99  1.01  ...  Christy Cunningham  3560325168603410
Sun4608          10.34  1.66  ...      Douglas Tucker  4478071379779230
Sun4458          21.01  3.50  ...      Travis Walters  6011812112971322
Sun5260          23.68  3.31  ...    Nathaniel Harris  4676137647685994
Sun2251          24.59  3.61  ...        Tonya Carter  4832732618637221

[5 rows x 10 columns]

#### Saisir une seule ligne

In [ ]:
# à partir d'un entier
df.iloc[0]

total_bill                       16.99
tip                               1.01
sex                             Female
smoker                              No
day                                Sun
time                            Dinner
size                                 2
price_per_person                  8.49
Payer Name          Christy Cunningham
CC Number             3560325168603410
Name: Sun2959, dtype: object

In [ ]:
# à partir d'un nom
df.loc['Sun2959']

total_bill                       16.99
tip                               1.01
sex                             Female
smoker                              No
day                                Sun
time                            Dinner
size                                 2
price_per_person                  8.49
Payer Name          Christy Cunningham
CC Number             3560325168603410
Name: Sun2959, dtype: object

#### Saisir plusieurs lignes

In [ ]:
df.iloc[0:4]

total_bill   tip  ...          Payer Name         CC Number
Payment ID                    ...                                      
Sun2959          16.99  1.01  ...  Christy Cunningham  3560325168603410
Sun4608          10.34  1.66  ...      Douglas Tucker  4478071379779230
Sun4458          21.01  3.50  ...      Travis Walters  6011812112971322
Sun5260          23.68  3.31  ...    Nathaniel Harris  4676137647685994

[4 rows x 10 columns]

In [ ]:
df.loc[['Sun2959','Sun5260']]

total_bill   tip  ...          Payer Name         CC Number
Payment ID                    ...                                      
Sun2959          16.99  1.01  ...  Christy Cunningham  3560325168603410
Sun5260          23.68  3.31  ...    Nathaniel Harris  4676137647685994

[2 rows x 10 columns]

#### Supprimer une ligne

Généralement, les jeux de données (datasets) sont suffisamment volumineux pour ne pas supprimer de lignes comme celle-ci, car nous ne connaîtrons pas l'emplacement de leur ligne pour une condition spécifique. Nous supprimons plutôt des lignes en fonction de conditions telles que les données manquantes ou les valeurs de colonne. La prochaine leçon abordera cette question de façon beaucoup plus détaillée.

In [ ]:
df.head()

total_bill   tip  ...          Payer Name         CC Number
Payment ID                    ...                                      
Sun2959          16.99  1.01  ...  Christy Cunningham  3560325168603410
Sun4608          10.34  1.66  ...      Douglas Tucker  4478071379779230
Sun4458          21.01  3.50  ...      Travis Walters  6011812112971322
Sun5260          23.68  3.31  ...    Nathaniel Harris  4676137647685994
Sun2251          24.59  3.61  ...        Tonya Carter  4832732618637221

[5 rows x 10 columns]

In [ ]:
df.drop('Sun2959',axis=0).head()

total_bill   tip  ...        Payer Name         CC Number
Payment ID                    ...                                    
Sun4608          10.34  1.66  ...    Douglas Tucker  4478071379779230
Sun4458          21.01  3.50  ...    Travis Walters  6011812112971322
Sun5260          23.68  3.31  ...  Nathaniel Harris  4676137647685994
Sun2251          24.59  3.61  ...      Tonya Carter  4832732618637221
Sun9679          25.29  4.71  ...        Erik Smith   213140353657882

[5 rows x 10 columns]

In [ ]:
# Erreur si vous avez un index étiqueté avec des chaines de nom !
# df.drop(0,axis=0).head()

#### Insérer une nouvelle ligne

Il est plutôt rare d'ajouter une ligne comme ceci. Habituellement, vous utilisez pd.concat() pour ajouter plusieurs lignes à la fois. Vous pouvez utiliser la méthode .append() avec une liste d'objets pd.Series(), mais vous ne nous verrez pas faire cela avec des données  réalistes issues du monde réel.

In [ ]:
one_row = df.iloc[0]

In [ ]:
one_row

total_bill                       16.99
tip                               1.01
sex                             Female
smoker                              No
day                                Sun
time                            Dinner
size                                 2
price_per_person                  8.49
Payer Name          Christy Cunningham
CC Number             3560325168603410
Name: Sun2959, dtype: object

In [ ]:
type(one_row)

pandas.core.series.Series

In [ ]:
df.tail()

total_bill   tip  ...       Payer Name         CC Number
Payment ID                    ...                                   
Sat2657          29.03  5.92  ...    Michael Avila  5296068606052842
Sat1766          27.18  2.00  ...   Monica Sanders  3506806155565404
Sat3880          22.67  2.00  ...       Keith Wong  6011891618747196
Sat17            17.82  1.75  ...     Dennis Dixon     4375220550950
Thur672          18.78  3.00  ...  Michelle Hardin  3511451626698139

[5 rows x 10 columns]

In [ ]:
df.append(one_row).tail()

total_bill   tip  ...          Payer Name         CC Number
Payment ID                    ...                                      
Sat1766          27.18  2.00  ...      Monica Sanders  3506806155565404
Sat3880          22.67  2.00  ...          Keith Wong  6011891618747196
Sat17            17.82  1.75  ...        Dennis Dixon     4375220550950
Thur672          18.78  3.00  ...     Michelle Hardin  3511451626698139
Sun2959          16.99  1.01  ...  Christy Cunningham  3560325168603410

[5 rows x 10 columns]

--------